File: csf_tz/csf_tz/page/salary_calculator/salary_calculator.js
frappe.pages["salary-calculator"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Salary Calculator",
		single_column: true,
	});
	frappe.breadcrumbs.add("HR");
	new SalaryCalculatorPage(page);
};

class SalaryCalculatorPage {
	constructor(page) {
		this.page = page;
		this.state = {
			salary_structure: "",
			employee: "",
			calculate_based_on: "Net Pay",
			gross_pay: 0,
			net_pay: 0,
		};
		this.structure_data = null; // { earnings, deductions, currency }
		this.active_keys = new Set(); // set of component keys currently active
		this.result = null;
		this.currency = "TZS";
		this._calc_timer = null;

		this.render_layout();
		this.create_fields();
	}

	// ── Layout ──────────────────────────────────────────────────────────

	render_layout() {
		this.page.main.addClass("salary-calculator-page").html(`
			<div class="sc-container">
				<div class="sc-left-panel">
					<div class="sc-card">
						<div class="sc-card-header">
							<h3 class="sc-card-title">Calculator Inputs</h3>
						</div>
						<div class="sc-card-body sc-fields"></div>
					</div>
					<div class="sc-card sc-components-card" style="display:none">
						<div class="sc-card-header">
							<h3 class="sc-card-title">Salary Components</h3>
							<div class="sc-header-right">
								<span class="sc-badge sc-count-badge">0 active</span>
								<button class="btn btn-xs btn-default sc-add-btn" style="display:none">+ Add</button>
							</div>
						</div>
						<div class="sc-card-body">
							<div class="sc-section">
								<div class="sc-section-label"><span class="sc-dot sc-dot-green"></span> Earnings</div>
								<div class="sc-list sc-earnings-list"></div>
							</div>
							<div class="sc-section">
								<div class="sc-section-label"><span class="sc-dot sc-dot-red"></span> Deductions</div>
								<div class="sc-list sc-deductions-list"></div>
							</div>
						</div>
					</div>
					<div class="sc-card sc-results-card" style="display:none">
						<div class="sc-card-header"><h3 class="sc-card-title">Calculation Results</h3></div>
						<div class="sc-card-body">
							<div class="sc-results-grid"></div>
							<div class="sc-results-breakdown"></div>
						</div>
					</div>
				</div>
				<div class="sc-right-panel">
					<div class="sc-card sc-preview-card">
						<div class="sc-card-header">
							<h3 class="sc-card-title">Salary Slip Preview</h3>
							<div class="sc-preview-actions"></div>
						</div>
						<div class="sc-card-body">
							<div class="sc-placeholder">
								<svg width="48" height="48" viewBox="0 0 24 24" fill="none" stroke="var(--gray-400)" stroke-width="1.5">
									<path d="M14 2H6a2 2 0 00-2 2v16a2 2 0 002 2h12a2 2 0 002-2V8z"/>
									<polyline points="14 2 14 8 20 8"/><line x1="16" y1="13" x2="8" y2="13"/>
									<line x1="16" y1="17" x2="8" y2="17"/>
								</svg>
								<p>Select a salary structure and enter an amount to begin</p>
							</div>
							<div class="sc-preview-content" style="display:none"></div>
						</div>
					</div>
				</div>
			</div>
		`);
	}

	// ── Fields ───────────────────────────────────────────────────────────

	create_fields() {
		const $fields = this.page.main.find(".sc-fields");

		this.fields = {};
		const defs = [
			{
				fieldtype: "Link", fieldname: "salary_structure", label: "Salary Structure",
				options: "Salary Structure", reqd: 1,
				get_query: () => ({ filters: { is_active: "Yes", docstatus: 1 } }),
				change: () => this.on_structure_change(),
			},
			{
				fieldtype: "Link", fieldname: "employee", label: "Employee",
				options: "Employee",
				get_query: () => ({ filters: { status: "Active" } }),
				change: () => { this.state.employee = this.val("employee"); this.refresh_preview(); },
			},
			{
				fieldtype: "Select", fieldname: "calculate_based_on", label: "Calculate Based On",
				options: ["Net Pay", "Gross Pay"], default: "Net Pay", reqd: 1,
				change: () => { this.state.calculate_based_on = this.val("calculate_based_on"); this.apply_field_states(); this.schedule_calc(); },
			},
		];

		for (const df of defs) {
			const $w = $('<div class="sc-field-wrap"></div>').appendTo($fields);
			this.fields[df.fieldname] = frappe.ui.form.make_control({ df, parent: $w, render_input: true });
		}
		this.fields.calculate_based_on.set_value("Net Pay");

		// Amount row
		const $row = $('<div class="sc-amount-row"></div>').appendTo($fields);
		for (const name of ["gross_pay", "net_pay"]) {
			const $w = $('<div class="sc-field-wrap sc-field-half"></div>').appendTo($row);
			this.fields[name] = frappe.ui.form.make_control({
				df: {
					fieldtype: "Currency", fieldname: name, precision: 0,
					label: name === "gross_pay" ? "Gross Pay" : "Net Pay",
					change: () => {
						this.state.gross_pay = flt(this.val("gross_pay"));
						this.state.net_pay = flt(this.val("net_pay"));
						this.schedule_calc();
					},
				},
				parent: $w, render_input: true,
			});
		}

		this.apply_field_states();
	}

	val(name) { return this.fields[name]?.get_value(); }

	apply_field_states() {
		const is_net = this.state.calculate_based_on === "Net Pay";
		for (const [name, disabled] of [["gross_pay", is_net], ["net_pay", !is_net]]) {
			const f = this.fields[name];
			if (f?.$input) {
				f.$input.prop("disabled", disabled);
				f.$wrapper.toggleClass("sc-field-disabled", disabled);
			}
		}
	}

	// ── Structure Change ────────────────────────────────────────────────

	on_structure_change() {
		const value = this.val("salary_structure");
		this.state.salary_structure = value;

		if (!value) {
			this.structure_data = null;
			this.active_keys = new Set();
			this.result = null;
			this.$(".sc-components-card, .sc-results-card").hide();
			this.show_placeholder();
			return;
		}

		frappe.xcall(
			"csf_tz.csf_tz.page.salary_calculator.salary_calculator.get_salary_structure_components",
			{ salary_structure: value },
		).then((data) => {
			this.structure_data = data;
			this.currency = data.currency || "TZS";

			// Auto-activate all visible components
			this.active_keys = new Set();
			for (const comp of [...data.earnings, ...data.deductions]) {
				if (!comp.statistical_component && !comp.do_not_include_in_total) {
					this.active_keys.add(comp.key);
				}
			}

			this.render_components();
			this.$(".sc-components-card").show();
			this.schedule_calc();
		});
	}

	// ── Components UI ───────────────────────────────────────────────────

	render_components() {
		const data = this.structure_data;
		if (!data) return;

		this.render_component_list(this.$(".sc-earnings-list"), data.earnings);
		this.render_component_list(this.$(".sc-deductions-list"), data.deductions);
		this.update_counts();
	}

	render_component_list($list, components) {
		$list.empty();
		let has_rows = false;

		for (const comp of components) {
			if (comp.statistical_component || comp.do_not_include_in_total) continue;
			if (!this.active_keys.has(comp.key)) continue;
			has_rows = true;
			$list.append(this.build_component_row(comp));
		}

		if (!has_rows) {
			$list.append('<div class="sc-empty-msg text-muted">None</div>');
		}
	}

	build_component_row(comp) {
		const is_base = this.is_base(comp);
		const hint = comp.formula
			? `Formula: ${comp.formula}`
			: comp.amount ? `Amount: ${comp.amount}` : "";
		const esc = frappe.utils.escape_html;

		const $row = $(`
			<div class="sc-comp-row" data-key="${comp.key}">
				<div class="sc-comp-info">
					<span class="sc-comp-name">${esc(comp.salary_component)}</span>
					${is_base ? '<span class="sc-tag">Base</span>' : ""}
					${hint ? `<span class="sc-comp-hint">${esc(hint)}</span>` : ""}
				</div>
				${!is_base ? '<button class="btn btn-xs sc-remove-btn" title="Remove">&times;</button>' : ""}
			</div>
		`);

		if (!is_base) {
			$row.find(".sc-remove-btn").on("click", () => {
				this.active_keys.delete(comp.key);
				this.render_components();
				this.schedule_calc();
			});
		}
		return $row;
	}

	is_base(comp) {
		const f = (comp.formula || "").replace(/\s/g, "").toLowerCase();
		return comp.abbr === "B" || (comp.salary_component || "").toLowerCase() === "basic" || f === "base";
	}

	get_inactive_components() {
		if (!this.structure_data) return [];
		return [...this.structure_data.earnings, ...this.structure_data.deductions].filter(
			(c) => !c.statistical_component && !c.do_not_include_in_total && !this.is_base(c) && !this.active_keys.has(c.key),
		);
	}

	update_counts() {
		this.$(".sc-count-badge").text(`${this.active_keys.size} active`);

		const inactive = this.get_inactive_components();
		const $btn = this.$(".sc-add-btn");
		if (inactive.length) {
			$btn.show().off("click").on("click", () => this.show_add_dialog());
		} else {
			$btn.hide();
		}
	}

	show_add_dialog() {
		const inactive = this.get_inactive_components();
		if (!inactive.length) return;

		const d = new frappe.ui.Dialog({
			title: __("Add Salary Component"),
			fields: [{
				fieldtype: "Select", fieldname: "key", label: "Component", reqd: 1,
				options: inactive.map((c) => ({ label: c.salary_component, value: c.key })),
			}],
			primary_action_label: __("Add"),
			primary_action: ({ key }) => {
				d.hide();
				this.active_keys.add(key);
				this.render_components();
				this.schedule_calc();
			},
		});
		d.show();
	}

	// ── Calculation ─────────────────────────────────────────────────────

	get_selected_names() {
		if (!this.structure_data) return [];
		const names = [];
		for (const comp of [...this.structure_data.earnings, ...this.structure_data.deductions]) {
			if (this.active_keys.has(comp.key) && !comp.statistical_component && !comp.do_not_include_in_total) {
				names.push(comp.salary_component);
			}
		}
		return names;
	}

	schedule_calc() {
		clearTimeout(this._calc_timer);
		this._calc_timer = setTimeout(() => this.run_calculation(), 300);
	}

	run_calculation() {
		const ss = this.state.salary_structure;
		const mode = this.state.calculate_based_on;
		if (!ss || !mode) return;

		const target = mode === "Net Pay" ? flt(this.val("net_pay")) : flt(this.val("gross_pay"));
		if (!target) {
			this.result = null;
			this.$(".sc-results-card").hide();
			this.show_placeholder();
			return;
		}

		frappe.xcall(
			"csf_tz.csf_tz.page.salary_calculator.salary_calculator.run_calculation",
			{
				salary_structure: ss,
				calculate_based_on: mode,
				gross_pay: this.state.gross_pay,
				net_pay: this.state.net_pay,
				selected_components: this.get_selected_names(),
			},
		).then((r) => {
			if (!r) return;
			this.result = r;

			// Fill the computed field without re-triggering calc
			if (mode === "Net Pay") {
				this.fields.gross_pay.set_value(r.gross_pay);
			} else {
				this.fields.net_pay.set_value(r.net_pay);
			}

			this.render_results(r);
			this.refresh_preview();
		}).catch((err) => {
			this.result = null;
			frappe.show_alert({ message: __("Calculation error: {0}", [err?.message || err]), indicator: "orange" }, 5);
		});
	}

	// ── Results ──────────────────────────────────────────────────────────

	render_results(r) {
		this.$(".sc-results-card").show();

		this.$(".sc-results-grid").html(`
			<div class="sc-stat"><div class="sc-stat-label">Base</div><div class="sc-stat-value">${this.fmt(r.base)}</div></div>
			<div class="sc-stat"><div class="sc-stat-label">Gross Pay</div><div class="sc-stat-value sc-text-green">${this.fmt(r.gross_pay)}</div></div>
			<div class="sc-stat"><div class="sc-stat-label">Deductions</div><div class="sc-stat-value sc-text-red">${this.fmt(r.total_deductions)}</div></div>
			<div class="sc-stat sc-stat-primary"><div class="sc-stat-label">Net Pay</div><div class="sc-stat-value">${this.fmt(r.net_pay)}</div></div>
		`);

		const comps = r.results || [];
		if (comps.length) {
			const rows = comps.map((c) =>
				`<tr><td>${frappe.utils.escape_html(c.salary_component)}</td><td class="text-right">${this.fmt(c.amount)}</td></tr>`
			).join("");
			this.$(".sc-results-breakdown").html(`
				<h5 class="sc-sub-title">Component Breakdown</h5>
				<table class="sc-table"><thead><tr><th>Component</th><th class="text-right">Amount</th></tr></thead>
				<tbody>${rows}</tbody></table>
			`);
		} else {
			this.$(".sc-results-breakdown").html("");
		}
	}

	// ── Preview ──────────────────────────────────────────────────────────

	refresh_preview() {
		const emp = this.state.employee;
		const ss = this.state.salary_structure;
		const r = this.result;

		if (!emp || !ss || !r) {
			this.show_placeholder();
			return;
		}

		const earnings = [], deductions = [];
		if (this.structure_data) {
			for (const comp of this.structure_data.earnings) {
				if (comp.statistical_component || comp.do_not_include_in_total) continue;
				if (!this.active_keys.has(comp.key)) continue;
				const match = (r.results || []).find((x) => x.salary_component === comp.salary_component);
				earnings.push({ salary_component: comp.salary_component, amount: match ? match.amount : (this.is_base(comp) ? r.base : 0) });
			}
			for (const comp of this.structure_data.deductions) {
				if (comp.statistical_component || comp.do_not_include_in_total) continue;
				if (!this.active_keys.has(comp.key)) continue;
				const match = (r.results || []).find((x) => x.salary_component === comp.salary_component);
				if (match) deductions.push({ salary_component: comp.salary_component, amount: match.amount });
			}
		}

		frappe.xcall(
			"csf_tz.csf_tz.page.salary_calculator.salary_calculator.get_salary_slip_preview",
			{ employee: emp, salary_structure: ss, base: r.base, gross_pay: r.gross_pay, net_pay: r.net_pay, earnings_data: earnings, deductions_data: deductions },
		).then((html) => {
			this.$(".sc-placeholder").hide();
			this.$(".sc-preview-content").html(html).show();
			this.render_preview_actions();
		});
	}

	show_placeholder() {
		this.$(".sc-placeholder").show();
		this.$(".sc-preview-content").hide();
		this.$(".sc-preview-actions").empty();
	}

	// ── Assignment ───────────────────────────────────────────────────────

	render_preview_actions() {
		const $a = this.$(".sc-preview-actions").empty();
		if (!this.state.employee || !this.result) return;

		$('<button class="btn btn-primary btn-sm">Create Assignment</button>')
			.appendTo($a)
			.on("click", () => this.show_assign_dialog());
	}

	show_assign_dialog() {
		const d = new frappe.ui.Dialog({
			title: __("Create Salary Structure Assignment"),
			fields: [
				{ fieldtype: "Link", fieldname: "employee", label: "Employee", options: "Employee", default: this.state.employee, read_only: 1 },
				{ fieldtype: "Link", fieldname: "salary_structure", label: "Salary Structure", options: "Salary Structure", default: this.state.salary_structure, read_only: 1 },
				{ fieldtype: "Currency", fieldname: "base", label: "Base Amount", default: this.result?.base || 0, precision: 0, description: "Calculated base salary" },
				{ fieldtype: "Date", fieldname: "from_date", label: "From Date", reqd: 1, default: frappe.datetime.get_today() },
			],
			primary_action_label: __("Create & Submit"),
			primary_action: (v) => {
				d.hide();
				frappe.xcall(
					"csf_tz.csf_tz.page.salary_calculator.salary_calculator.create_salary_structure_assignment",
					{ employee: v.employee, salary_structure: v.salary_structure, from_date: v.from_date, base: v.base },
				).then((name) => {
					frappe.show_alert({
						message: __("Assignment {0} created", [`<a href="/app/salary-structure-assignment/${name}">${name}</a>`]),
						indicator: "green",
					}, 7);
				});
			},
		});
		d.show();
	}

	// ── Helpers ──────────────────────────────────────────────────────────

	$(selector) { return this.page.main.find(selector); }

	fmt(amount) {
		return format_number(flt(amount), null, this.currency === "TZS" ? 0 : 2);
	}
}

File: csf_tz/csf_tz/page/salary_calculator/salary_calculator.py
"""Salary Calculator page backend.

Provides a resilient calculation engine that handles salary structures whose
formulas/conditions reference external variables (absent_days, advance, etc.)
by silently setting those components to zero instead of crashing.
"""
from __future__ import annotations

import json
import re

import frappe
from frappe import _
from frappe.utils import cint, cstr, flt


# ── Constants ────────────────────────────────────────────────────────

_SAFE_GLOBALS = {"int": int, "float": float, "round": round, "abs": abs, "min": min, "max": max, "flt": flt}
_BOOLEAN_PREFIXES = ("custom_deduct_", "deduct_", "include_", "custom_")


# ── Public API ───────────────────────────────────────────────────────

@frappe.whitelist()
def get_salary_structure_components(salary_structure):
    """Return earnings/deductions for a salary structure with unique keys."""
    structure = frappe.get_cached_doc("Salary Structure", salary_structure)

    def serialize(rows, prefix):
        return [
            {
                "key": f"{prefix}-{i}",
                "salary_component": row.salary_component,
                "abbr": row.abbr,
                "amount": flt(row.amount),
                "amount_based_on_formula": row.amount_based_on_formula,
                "formula": row.formula or "",
                "condition": row.condition or "",
                "do_not_include_in_total": row.do_not_include_in_total,
                "statistical_component": row.statistical_component,
            }
            for i, row in enumerate(rows)
        ]

    return {
        "earnings": serialize(structure.earnings, "E"),
        "deductions": serialize(structure.deductions, "D"),
        "currency": structure.currency or "",
    }


@frappe.whitelist()
def run_calculation(salary_structure, calculate_based_on, gross_pay=0, net_pay=0, selected_components=None):
    """Run salary calculation with resilient formula evaluation."""
    if isinstance(selected_components, str):
        selected_components = json.loads(selected_components)
    selected = list(dict.fromkeys(selected_components or []))

    structure = frappe.get_cached_doc("Salary Structure", salary_structure)
    precision = 0 if cstr(structure.currency) == "TZS" else 2
    target_field = "gross_pay" if calculate_based_on == "Gross Pay" else "net_pay"
    target_amount = flt({"gross_pay": gross_pay, "net_pay": net_pay}[target_field], precision)

    if target_amount <= 0:
        return _empty_result(precision)

    return _solve(structure, target_field, target_amount, precision, selected)


@frappe.whitelist()
def get_salary_slip_preview(employee, salary_structure, base, gross_pay, net_pay, earnings_data, deductions_data):
    """Render a salary slip preview from calculated data."""
    if isinstance(earnings_data, str):
        earnings_data = json.loads(earnings_data)
    if isinstance(deductions_data, str):
        deductions_data = json.loads(deductions_data)

    emp = frappe.get_cached_doc("Employee", employee)
    currency = frappe.get_cached_value("Salary Structure", salary_structure, "currency") or "TZS"
    fmt = (lambda a: format(flt(a), ",.0f")) if currency == "TZS" else (lambda a: format(flt(a), ",.2f"))

    return frappe.render_template(
        "csf_tz/csf_tz/page/salary_calculator/salary_slip_preview.html",
        {
            "employee": employee,
            "employee_name": emp.employee_name,
            "department": emp.department or "",
            "designation": emp.designation or "",
            "company": emp.company,
            "salary_structure": salary_structure,
            "currency": currency,
            "base": flt(base),
            "gross_pay": flt(gross_pay),
            "net_pay": flt(net_pay),
            "total_earning": sum(flt(e.get("amount")) for e in earnings_data),
            "total_deduction": sum(flt(d.get("amount")) for d in deductions_data),
            "earnings": earnings_data,
            "deductions": deductions_data,
            "format_amount": fmt,
        },
    )


@frappe.whitelist()
def create_salary_structure_assignment(employee, salary_structure, from_date, base=0):
    """Create and submit a Salary Structure Assignment."""
    if frappe.db.exists(
        "Salary Structure Assignment",
        {"employee": employee, "salary_structure": salary_structure, "from_date": from_date, "docstatus": ["!=", 2]},
    ):
        frappe.throw(
            _("Salary Structure Assignment already exists for {0} with {1} from {2}").format(
                frappe.bold(employee), frappe.bold(salary_structure), frappe.bold(from_date),
            )
        )

    ssa = frappe.new_doc("Salary Structure Assignment")
    ssa.employee = employee
    ssa.salary_structure = salary_structure
    ssa.from_date = from_date
    ssa.base = flt(base)
    ssa.company = frappe.get_cached_value("Employee", employee, "company")
    ssa.save()
    ssa.submit()
    return ssa.name


# ── Resilient Calculation Engine ─────────────────────────────────────

def _empty_result(precision=0):
    return {"base": 0, "gross_pay": 0, "net_pay": 0, "total_deductions": 0, "results": []}


def _solve(structure, target_field, target_amount, precision, selected):
    """Binary search for the base amount that produces the target gross/net pay."""
    tol = 1 if precision == 0 else 0.05
    lo, hi = 0.0, max(target_amount, 1)
    best, best_diff = None, None

    # Phase 1: find upper bound
    for i in range(20):
        r = _calc(structure, hi, precision, selected)
        val = flt(r[target_field], precision)
        if val >= target_amount:
            best, best_diff = r, abs(val - target_amount)
            break
        # If doubling base doesn't change the target value, formulas aren't responding
        if i > 2 and val == 0:
            return _empty_result(precision)
        hi *= 2
    else:
        best = r
        best_diff = abs(flt(r[target_field], precision) - target_amount)

    # Phase 2: binary search
    for _ in range(60):
        mid = (lo + hi) / 2
        r = _calc(structure, mid, precision, selected)
        diff = flt(r[target_field], precision) - target_amount
        if abs(diff) < (best_diff or float("inf")):
            best, best_diff = r, abs(diff)
        if abs(diff) <= tol:
            break
        if diff < 0:
            lo = mid
        else:
            hi = mid

    # Phase 3: refine around best integer
    base_int = int(round(flt(best["base"], precision)))
    for offset in range(-3, 4):
        cand = base_int + offset
        if cand < 0:
            continue
        r = _calc(structure, cand, precision, selected)
        d = abs(flt(r[target_field], precision) - target_amount)
        if d < best_diff or (d == best_diff and flt(r[target_field], precision) >= flt(best[target_field], precision)):
            best, best_diff = r, d

    return best


def _calc(structure, base_amount, precision, selected):
    """Calculate salary from a base amount with resilient formula evaluation."""
    rows = _init_rows(structure, precision)
    gross_pay = net_pay = total_ded = 0.0
    prev = None

    for _ in range(10):
        ctx = {"base": flt(base_amount, precision), "gross_pay": gross_pay, "net_pay": net_pay, "total_deductions": total_ded}
        _add_row_ctx(ctx, rows["earnings"] + rows["deductions"], precision)
        _add_flag_ctx(ctx, rows["earnings"] + rows["deductions"], selected)

        earnings = _eval_rows(rows["earnings"], ctx, precision, selected)
        gross_pay = flt(sum(r.amount for r in earnings if _incl_gross(r, selected)), precision)
        ctx.update({"gross_pay": gross_pay})
        _add_row_ctx(ctx, earnings, precision)

        deductions = _eval_rows(rows["deductions"], ctx, precision, selected)
        total_ded = flt(sum(r.amount for r in deductions if _incl_ded(r, selected)), precision)
        net_pay = flt(gross_pay - total_ded, precision)

        state = (gross_pay, total_ded, net_pay, tuple(r.amount for r in earnings + deductions))
        rows = {"earnings": earnings, "deductions": deductions}
        if state == prev:
            break
        prev = state

    all_rows = earnings + deductions
    results = _build_results(selected, all_rows, precision)
    return {
        "base": flt(base_amount, precision),
        "gross_pay": gross_pay,
        "net_pay": net_pay,
        "total_deductions": total_ded,
        "results": results,
    }


def _init_rows(structure, precision):
    def norm(row, ctype):
        return frappe._dict(
            salary_component=cstr(row.salary_component), abbr=cstr(row.abbr), component_type=ctype,
            condition=cstr(row.condition), formula=cstr(row.formula),
            amount=flt(row.amount, precision), amount_based_on_formula=cint(row.amount_based_on_formula),
            do_not_include_in_total=cint(row.do_not_include_in_total), statistical_component=cint(row.statistical_component),
        )
    return {
        "earnings": [norm(r, "Earning") for r in structure.earnings],
        "deductions": [norm(r, "Deduction") for r in structure.deductions],
    }


def _add_row_ctx(ctx, rows, precision):
    for r in rows:
        if r.abbr:
            ctx[r.abbr] = flt(r.amount, precision)
            ctx[f"{r.abbr}_amount"] = flt(r.amount, precision)


def _add_flag_ctx(ctx, rows, selected):
    """Infer boolean/numeric flags from formulas and conditions."""
    exprs = [r.condition for r in rows if r.condition] + [r.formula for r in rows if r.formula]
    tokens = {
        tok for expr in exprs
        for tok in re.findall(r"\b[A-Za-z_][A-Za-z0-9_]*\b", expr)
        if any(tok.startswith(p) for p in _BOOLEAN_PREFIXES)
    }
    for tok in tokens:
        if tok.startswith(_BOOLEAN_PREFIXES):
            kw = tok
            for p in _BOOLEAN_PREFIXES:
                if kw.startswith(p):
                    kw = kw[len(p):]
                    break
            norm = kw.replace("_", "").lower()
            sel_text = " ".join(selected).replace(" ", "").replace("_", "").lower()
            ctx.setdefault(tok, 1 if norm and norm in sel_text else 0)
        else:
            ctx.setdefault(tok, 0)


def _eval_rows(rows, base_ctx, precision, selected):
    ctx = frappe._dict(base_ctx.copy())
    result = []
    for row in rows:
        amt = row.amount if not row.amount_based_on_formula else 0
        if not _should_eval(row, selected):
            amt = 0
        elif row.condition and not _safe_condition(row.condition, ctx):
            amt = 0
        elif row.amount_based_on_formula and row.formula:
            amt = _safe_formula(row.formula, ctx, precision)

        row.amount = flt(amt, precision)
        result.append(row)
        if row.abbr:
            ctx[row.abbr] = row.amount
            ctx[f"{row.abbr}_amount"] = row.amount
    return result


def _safe_formula(formula, ctx, precision):
    """Evaluate a formula, returning 0 on any error."""
    try:
        return flt(frappe.safe_eval(cstr(formula).strip(), eval_globals=_SAFE_GLOBALS, eval_locals=ctx), precision)
    except Exception:
        return 0


def _safe_condition(condition, ctx):
    """Evaluate a condition, returning False on any error."""
    try:
        return cint(frappe.safe_eval(cstr(condition).strip(), eval_globals=_SAFE_GLOBALS, eval_locals=ctx))
    except Exception:
        return False


def _is_base(row):
    f = cstr(row.formula).replace(" ", "").lower()
    return row.abbr == "B" or cstr(row.salary_component).lower() == "basic" or f == "base"


def _should_eval(row, selected):
    if row.statistical_component or _is_base(row):
        return True
    return cstr(row.salary_component) in set(selected)


def _incl_gross(row, selected):
    if row.component_type != "Earning" or row.do_not_include_in_total or row.statistical_component:
        return False
    return _is_base(row) or cstr(row.salary_component) in set(selected)


def _incl_ded(row, selected):
    if row.component_type != "Deduction" or row.do_not_include_in_total or row.statistical_component:
        return False
    return cstr(row.salary_component) in set(selected)


def _build_results(selected, all_rows, precision):
    amounts = {}
    for r in all_rows:
        name = cstr(r.salary_component)
        if name:
            amounts[name] = flt(amounts.get(name, 0) + flt(r.amount, precision), precision)
    return [{"salary_component": c, "amount": flt(amounts.get(c, 0), precision)} for c in selected]
